napari/_app_model/injection/_providers.py

```diff
@@ -1,10 +1,22 @@
+"""Non-Qt providers.
+
+Qt providers can be found in `napari/_qt/_qapp_model/injection/_qproviders.py`.
+
+Because `_provide_viewer` needs `_QtMainWindow` (otherwise returns `None`)
+tests are in `napari/_tests/test_providers.py`, which are not run in headless mode.
+"""
+
 from typing import Optional
 
 from napari import components, layers, viewer
+from napari.utils._proxies import PublicOnlyProxy
 
 
 def _provide_viewer() -> Optional[viewer.Viewer]:
-    return viewer.current_viewer()
+    """Provide `PublicOnlyProxy` (allows internal napari access) of current viewer."""
+    if current_viewer := viewer.current_viewer():
+        return PublicOnlyProxy(current_viewer)
+    return current_viewer
```

Comment:
```suggestion
    return None
```
I prefer to explicitly return None